### PR TITLE
[5064] RTE dialogs - data source issues

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/5064

### CODE REVIEW

File: ui/app/src/models/ContentType.ts [hunk 1]
```diff
@@ -46,7 +46,9 @@ export type ValidationKeys =
   | 'minValue'
   | 'maxValue'
   | 'dropTargetsNotFound'
-  | 'registerNotFound';
+  | 'registerNotFound'
+  | 'imgRepositoryUpload'
+  | 'imgDesktopUpload';
```
Comment (rart) [inline ui/app/src/models/ContentType.ts]:
```suggestion
  | 'allowImagesFromRepo'
  | 'allowImageUpload';
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/utils/path.ts [hunk 2]
```diff
@@ -245,3 +247,63 @@ export function getItemTemplatePath(item: DetailedItem, contentTypes: LookupTabl
...
 }
+
+export function processPathMacros(path: string, model: any | ContentInstance, useUUID?: boolean): string {
+  const descriptor = model.craftercms?.descriptor ?? parseDescriptor(model);
+  let processedPath = path;
+
+  if (processedPath.indexOf('{objectId}') !== -1) {
+    if (useUUID) {
+      // TODO: generateUUID function
+      // processedPath = processedPath.replace('{objectId}', CStudioAuthoring.Utils.generateUUID());
```
Comment (jvega190) [inline ui/app/src/utils/path.ts]: @rart need to discuss this utility

File: static-assets/components/cstudio-common/common-api.js [hunk 3]
```diff
@@ -2816,60 +2816,7 @@ var nodeOpen = false,
...
-        }
-
-        if (path.indexOf('{month}') != -1) {
-          path = path.replace('{month}', ('0' + (currentDate.getMonth() + 1)).slice(-2));
-        }
-
-        const fullParentPath =
-          CStudioAuthoring.Utils.getQueryParameterByName('path') !== ''
-            ? CStudioAuthoring.Utils.getQueryParameterByName('path')
-            : CStudioAuthoring.Utils.getQueryParameterByName('parentPath');
```
Comment (rart) [inline static-assets/components/cstudio-common/common-api.js:2848]: All these should become arguments to the new function. Don't read URL there.

File: ui/app/src/components/DataSourcesActionsList/DataSourcesActionsList.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+export default function DataSourcesActionsList(props: DataSourcesActionsListProps) {
+  const { show, rect, items, onClose } = props;
+  const sx = getStyles();
+
+  return (
+    <>
+      {items.length > 0 && (
+        <Dialog
```
Comment (rart) [inline ui/app/src/components/DataSourcesActionsList/DataSourcesActionsList.tsx]: Use `Menu`

File: ui/app/src/components/DataSourcesActionsList/DataSourcesActionsList.tsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            style: {
+              top: rect.top,
+              left: rect.left
+            }
+          }}
+          BackdropProps={{ sx: sx.backdrop }}
+        >
+          <List dense={true}>
+            {items.map((item, index) => (
+              <ListItem key={index} disableGutters>
```
Comment (rart) [inline ui/app/src/components/DataSourcesActionsList/DataSourcesActionsList.tsx]: Use `MenuItem`

File: ui/app/src/state/actions/dialogs.ts [hunk 6]
```diff
@@ -269,3 +269,10 @@ export const updateWidgetDialog = /*#__PURE__*/ createAction<Partial<WidgetDialo
 // region Show Keyboard Shortcuts Dialog
 export const showKeyboardShortcutsDialog = /*#__PURE__*/ createAction('SHOW_KEYBOARD_SHORTCUTS_DIALOG');
 // endregion
+
+// region Show Dialogs From DataSources
+export const showRtePickerActions = /*#__PURE__*/ createAction('SHOW_RTE_PICKER_ACTIONS');
+// endregion
+// region Rte Picker Action Result
```
Comment (rart) [inline ui/app/src/state/actions/dialogs.ts:277]: unclosed region

File: ui/app/src/utils/path.ts [hunk 7]
```diff
@@ -245,3 +247,63 @@ export function getItemTemplatePath(item: DetailedItem, contentTypes: LookupTabl
...
+
+  if (processedPath.indexOf('{mm}') !== -1) {
+    processedPath = processedPath.replace('{mm}', ('0' + (currentDate.getMonth() + 1)).slice(-2));
+  }
+
+  if (processedPath.indexOf('{dd}') !== -1) {
+    processedPath = processedPath.replace('{dd}', ('0' + currentDate.getDate()).slice(-2));
+  }
+
+  const urlParams = new URLSearchParams(window.location.search);
```
Comment (rart) [inline ui/app/src/utils/path.ts]: Don't read location or anything other than own arguments on this function. This function should be pure.

File: ui/app/src/utils/path.ts [hunk 8]
```diff
@@ -245,3 +247,63 @@ export function getItemTemplatePath(item: DetailedItem, contentTypes: LookupTabl
 export function getControllerPath(type: SystemType): string {
   return `/scripts/${type === 'page' ? 'pages' : 'components'}`;
 }
+
+export function processPathMacros(path: string, model: any | ContentInstance, useUUID?: boolean): string {
```
Comment (rart) [inline ui/app/src/utils/path.ts]:
```suggestion
export function processPathMacros(dependencies: { define all here }): string {
```

File: ui/app/src/utils/path.ts (hunk 2 above)
Comment (rart) [reply ui/app/src/utils/path.ts]: We have a `uuid` lib. 
```js
import { v4 as uuid } from 'uuid';
id = uuid();
```

File: ui/guest/src/controls/rte.ts [hunk 9]
```diff
@@ -114,6 +116,42 @@ export function initTinyMCE(
...
+          datasources,
+          model,
+          type: meta.filetype,
+          btnRect: browseBtn.getBoundingClientRect()
+        }
+      });
+
+      message$
+        .pipe(
+          filter((e) => e.type === rtePickerActionResult.type),
```
Comment (rart) [inline ui/guest/src/controls/rte.ts:147]: Make sure you're sending both ok and cancel actions so this observable dies with either result.

**Review by rart — CHANGES_REQUESTED** [review]

**Comment by rart** [comment]
Conflicts

**Review by rart — APPROVED** [review]

**Pull request merged**